aggregation.go

```diff
@@ -1 +1,281 @@
+	for _, aggName := range aggNameSlice {
+		colNameGroupBy[aggName] = 1
+	}
+	if len(colNameGroupBy) == 0 {
+		return nil
+	}
+	for _, colNameStr := range colNameSlice {
+		if _, exist := colNameGroupBy[colNameStr]; !exist {
+			err = fmt.Errorf(`esql: select column %v that not in group by`, colNameStr)
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *ESql) getAggOrderBy(orderBy sqlparser.OrderBy) ([]string, []string, []string, []string, map[string]int, error) {
+	var aggNameSlice, aggTargetSlice, aggDirSlice, aggTagSlice []string
+	aggTagDirSet := make(map[string]string) // tag -> asc / desc
+	aggTagSet := make(map[string]int)       // tag -> offset, for compatiblity checking
+
+	for _, orderExpr := range orderBy {
+		aggStr := strings.Trim(sqlparser.String(orderExpr.Expr), "`")
+		if strings.ContainsAny(aggStr, "()") {
+			// TODO: do more sanity checks, like prevent order the same target with different directions
+			// eg: COUNT(colA) -> count_colA
+			aggStr = strings.Trim(aggStr, ")")
+			strParts := strings.Split(aggStr, "(")
+			aggNameStr := strings.ToLower(strParts[0])
+			aggTargetStr := strParts[1]
+			var aggTagStr string
+			aggStr = strings.ToLower(aggStr)
+			switch aggNameStr {
+			case "count":
+				if !strings.Contains(aggStr, "distinct") {
+					aggTagStr = "_count"
+				} else {
+					aggTargetParts := strings.Split(aggTargetStr, " ")
+					if len(aggTargetParts) == 2 && strings.ToLower(aggTargetParts[0]) == "distinct" {
+						aggTagStr = aggNameStr + "_distinct_" + aggTargetParts[1]
+					} else {
+						err := fmt.Errorf(`esql: order by aggregation function %v input err`, aggStr+")")
+						return nil, nil, nil, nil, nil, err
+					}
+				}
+			case "sum", "min", "max", "avg":
+				if strings.Contains(aggStr, "distinct") {
+					err := fmt.Errorf(`esql: order by aggregation function %v w/ DISTINCT not supported`, aggNameStr)
+					return nil, nil, nil, nil, nil, err
+				}
+				aggTagStr = aggNameStr + "_" + aggTargetStr
+			default:
+				err := fmt.Errorf(`esql: order by aggregation function %v not supported`, aggNameStr)
+				return nil, nil, nil, nil, nil, err
+			}
+			if dir, exist := aggTagDirSet[aggTagStr]; exist {
+				if dir != orderExpr.Direction {
+					err := fmt.Errorf(`esql: order by aggregation direction conflict`)
+					return nil, nil, nil, nil, nil, err
+				}
+				continue
+			}
+			aggTagDirSet[aggTagStr] = orderExpr.Direction
+			aggTagSet[aggTagStr] = len(aggTagSet)
+			aggNameSlice = append(aggNameSlice, aggNameStr)
+			aggTargetSlice = append(aggTargetSlice, aggTargetStr)
+			aggTagSlice = append(aggTagSlice, aggTagStr)
+			aggDirSlice = append(aggDirSlice, orderExpr.Direction)
+		}
+	}
+	return aggNameSlice, aggTargetSlice, aggTagSlice, aggDirSlice, aggTagSet, nil
+}
+
+func (e *ESql) getAggSelect(exprs []*sqlparser.FuncExpr) ([]string, []string, []string, map[string]int, error) {
```

Comment:
Too many return types make it hard to understand. 
consider make return a struct, or at least add comments on what are returns.
Also might rename aggTagSet to be more precise, currently it confuse me what is it.

Reply:
I added some comments. Probably it is still not clear. Maybe we can briefly discuss it later.